File: CP/CF/CF_round_677_div3_alt/6.cpp
#include<bits/stdc++.h>
#define endl "\n"
using namespace std;
int n,m,k;
int dp[71][71][71][71], mat[71][71];

int rec(int i, int j, int count, int rem){
    if (i>=n) return rem==0 ? 0 : INT_MIN;
    if (j>=m) return rec(i+1,0,0,rem);
    int &res = dp[i][j][count][rem];
    if(res!=-1) return res;
    res = max(rec(i+1,0,0,rem), rec(i,j+1,count, rem));
    if(count<m/2){
        res = max(res, mat[i][j]+rec(i,j+1, count+1, (rem+mat[i][j])%k));
    }
    return res;
}

void solve(){
    cin>>n>>m>>k;
    for(int i =0; i<n; ++i)
        for(int j =0; j<m; ++j)
            cin>>mat[i][j];
    memset(dp,-1,sizeof(dp));
    cout<<rec(0,0,0,0)<<endl;
}

int main(){
    // freopen("in.txt", "r", stdin);
    int t = 1;
    // cin>>t;
    while(t-->0) solve();
    return 0;
}

File: CP/CF/CF_round_677_div3_alt/6.py
import os, sys, bisect
from collections import defaultdict, Counter, deque;
from functools import lru_cache   #use @lru_cache(None)
if os.path.exists('in.txt'): sys.stdin=open('in.txt','r')
if os.path.exists('out.txt'): sys.stdout=open('out.txt', 'w')
#
input = lambda: sys.stdin.readline().strip()
imap = lambda: map(int,input().split()); ilist = lambda: list(imap())
#------------------------------------------------------------------
sys.setrecursionlimit(10**6)
mod = int(1e9+7)

n,m,k = map(int,input().split())
mat = []
for i in range(n):
    mat.append(ilist())

@lru_cache(None)
def rec(i, j, count, rem):
    if i>=n:
        if rem==0: return 0
        return -float("inf")
    if j>=m:
        return rec(i+1, 0, 0, rem)
    res = max(rec(i+1, 0, 0,rem), rec(i, j+1, count,rem))
    if count<m//2:
        res = max(res, mat[i][j]+ rec(i, j+1, count+1, (rem+mat[i][j])%k))
        res = max(res, mat[i][j]+ rec(i+1, 0, 0, (rem+mat[i][j])%k))
    return res
print(rec(0,0,0,0))

